File: packages/canvas/src-native/canvas-ios/TNSCompositeOperationType.swift
//
//  TNSCompositeOperationType.swift
//
//  Created by Osei Fortune on 7/17/19.
//  Copyright © 2019 Osei Fortune. All rights reserved.
//

import Foundation

@objc(TNSCompositeOperationType)
public enum TNSCompositeOperationType: Int, RawRepresentable {
    case SourceOver
    case SourceIn
    case SourceOut
    case SourceAtop
    case DestinationOver
    case DestinationIn
    case DestinationOut
    case DestinationAtop
    case Lighter
    case Copy
    case Xor
    case Multiply
    case Screen
    case Overlay
    case Darken
    case Lighten
    case ColorDodge
    case ColorBurn
    case HardLight
    case SoftLight
    case Difference
    case Exclusion
    case Hue
    case Saturation
    case Color
    case Luminosity
    
    public typealias RawValue = UInt32
    
    
    public var rawValue: RawValue {
        switch self {
        case .SourceOver:
            return 0
        case .SourceIn:
            return 1
        case .SourceOut:
            return 2
        case .SourceAtop:
            return 3
        case .DestinationOver:
            return 4
        case .DestinationIn:
            return 5
        case .DestinationOut:
            return 6
        case  .DestinationAtop:
            return 7
        case .Lighter:
            return 8
        case .Copy:
            return 9
        case .Xor:
            return 10
        case .Multiply:
            return 11
        case .Screen:
            return 12
        case .Overlay:
            return 13
        case .Darken:
            return 14
        case .Lighten:
            return 15
        case .ColorDodge:
            return 16
        case .ColorBurn:
            return 17
        case .HardLight:
            return 18
        case .SoftLight:
            return 19
        case .Difference:
            return 20
        case .Exclusion:
            return 21
        case .Hue:
            return 22
        case .Saturation:
            return 23
        case .Color:
            return 24
        case .Luminosity:
            return 25
        }
    }
    
    public init?(rawValue: RawValue) {
        switch rawValue {
        case 0:
            self = .SourceOver
        case 1:
            self = .SourceIn
        case 2:
            self = .SourceOut
        case 3:
            self = .SourceAtop
        case 4:
            self = .DestinationOver
        case 5:
            self = .DestinationIn
        case 6:
            self = .DestinationOut
        case 7:
            self = .DestinationAtop
        case 8:
            self = .Lighter
        case 9:
            self = .Copy
        case 10:
            self = .Multiply
        case 11:
            self = .Xor
        case 12:
            self = .Screen
        case 13:
            self = .Overlay
        case 14:
            self = .Darken
        case 15:
            self = .Lighten
        case 16:
            self = .ColorDodge
        case 17:
            self = .ColorBurn
        case 18:
            self = .HardLight
        case 19:
            self = .SoftLight
        case 20:
            self = .Difference
        case 21:
            self = .Exclusion
        case 22:
            self = .Hue
        case 23:
            self = .Saturation
        case 24:
            self = .Luminosity
        default:
            return nil
        }
    }
    
    
    public init?(string: String) {
        switch string {
        case "source-over":
            self = .SourceOver
        case "source-in":
            self = .SourceIn
        case "source-out":
            self = .SourceOut
        case "source-atop":
            self = .SourceAtop
        case "destination-over":
            self = .DestinationOver
        case "destination-in":
            self = .DestinationIn
        case "destination-out":
            self = .DestinationOut
        case  "destination-atop":
            self = .DestinationAtop
        case "lighter":
            self = .Lighter
        case "copy":
            self = .Copy
        case "multiply":
            self = .Multiply
        case "xor":
            self = .Xor
        case "screen":
            self = .Screen
        case "overlay":
            self = .Overlay
        case "darken":
            self = .Darken
        case "lighten":
            self = .Lighten
        case "color-dodge":
            self = .ColorDodge
        case  "color-burn":
            self = .ColorBurn
        case "hard-light":
            self = .HardLight
        case "soft-light":
            self = .SoftLight
        case  "difference":
            self = .Difference
        case  "exclusion":
            self = .Exclusion
        case "hue":
            self = .Hue
        case "saturation":
            self = .Saturation
        case "luminosity":
            self = .Luminosity
        default:
            return nil
        }
    }
}

File: packages/canvas/src-native/canvas-android/canvas/src/main/java/org/nativescript/canvas/TNSCompositeOperationType.kt
package org.nativescript.canvas

/**
 * Created by triniwiz on 2019-07-13
 */
enum class TNSCompositeOperationType(private var type: String, private var value: Int) {
	SourceOver("source-over", 0), SourceIn("source-in", 1), SourceOut(
		"source-out",
		2
	),
	SourceAtop("source-atop", 3), DestinationOver(
		"destination-over", 4
	),
	DestinationIn("destination-in", 5), DestinationOut(
		"destination-out",
		6
	),
	DestinationAtop("destination-atop", 7), Lighter(
		"lighter", 8
	),
	Copy("copy", 9), Xor("xor", 10), Multiply("multiply", 11), Screen(
		"screen",
		12
	),
	Overlay("overlay", 13), Darken("darken", 14), Lighten(
		"lighten", 15
	),
	ColorDodge("color-dodge", 16), ColorBurn("color-burn", 17), HardLight(
		"hard-light",
		18
	),
	SoftLight("soft-light", 19), Difference(
		"difference", 20
	),
	Exclusion("exclusion", 21), Hue("hue", 22), Saturation("saturation", 23), Color(
		"color",
		24
	),
	Luminosity("luminosity", 25);

	override fun toString(): String {
		return type
	}

	fun toNative(): Int {
		return value
	}

	internal var isError = false

	companion object {
		fun fromNative(value: Int): TNSCompositeOperationType {
			return when (value) {
				0 -> SourceOver
				1 -> SourceIn
				2 -> SourceOut
				3 -> SourceAtop
				4 -> DestinationOver
				5 -> DestinationIn
				6 -> DestinationOut
				7 -> DestinationAtop
				8 -> Lighter
				9 -> Copy
				10 -> Multiply
				11 -> Xor
				12 -> Screen
				13 -> Overlay
				14 -> Darken
				15 -> Lighten
				16 -> ColorDodge
				17 -> ColorBurn
				18 -> HardLight
				19 -> SoftLight
				20 -> Difference
				21 -> Exclusion
				22 -> Hue
				23 -> Saturation
				24 -> Luminosity
				else -> SourceOver.apply { isError = true }
			}
		}
	}

}
